src/krb5_plugin/common/radius_kdcpreauth.c

```diff
@@ -0,0 +1,609 @@
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <limits.h>
+#include <krad.h>
+#include <krb5/kdcpreauth_plugin.h>
+
+#include "radius_kdcpreauth.h"
+#include "util/util.h"
+
+krb5_error_code
+sss_radiuskdc_init(const char *plugin_name,
+                   krb5_context kctx,
+                   krb5_kdcpreauth_moddata *_moddata,
+                   const char **_realmnames)
+{
+    struct sss_radiuskdc_state *state;
+
+    state = malloc(sizeof(struct sss_radiuskdc_state));
+    if (state == NULL) {
+        return ENOMEM;
+    }
+
+    state->plugin_name = plugin_name;
+
+    /* IPA is the only consumer so far so it is fine to hardcode the values. */
+    state->server = KRB5_KDC_RUNDIR "/DEFAULT.socket";
+    state->secret = "";
+    state->timeout = 5 * 1000;
+    state->retries = 3;
+
+    *_moddata = (krb5_kdcpreauth_moddata)state;
+
+    return 0;
+}
+
+void
+sss_radiuskdc_fini(krb5_context kctx,
+                   krb5_kdcpreauth_moddata moddata)
+{
+    struct sss_radiuskdc_state *state;
+
+    state = (struct sss_radiuskdc_state *)moddata;
+
+    if (state == NULL) {
+        return;
+    }
+
+    free(state);
+}
+
+int
+sss_radiuskdc_flags(krb5_context kctx,
+                    krb5_preauthtype pa_type)
+{
+    return PA_REPLACES_KEY;
+}
+
+krb5_error_code
+sss_radiuskdc_return_padata(krb5_context kctx,
+                            krb5_pa_data *padata,
+                            krb5_data *req_pkt,
+                            krb5_kdc_req *request,
+                            krb5_kdc_rep *reply,
+                            krb5_keyblock *encrypting_key,
+                            krb5_pa_data **send_pa_out,
+                            krb5_kdcpreauth_callbacks cb,
+                            krb5_kdcpreauth_rock rock,
+                            krb5_kdcpreauth_moddata moddata,
+                            krb5_kdcpreauth_modreq modreq)
+{
+    struct sss_radiuskdc_state *state;
+    krb5_keyblock *armor_key;
+    bool *result;
+
+    state = (struct sss_radiuskdc_state *)moddata;
+    result = (bool *)modreq;
+
+    /* This should not happen. */
+    if (state == NULL) {
+        return EINVAL;
+    }
+
+    /* Verification was not successful. Do not replace the key. */
+    if (result == NULL || *result == false) {
+        return 0;
+    }
+
+    /* Unexpected padata. Return error. */
+    if (padata->length != 0) {
```

Comment:
Hi,

this makes sense for IdP where we do not expect a reply. But for passkey there will be padata in the response. So I think this check should be removed here and maybe added back to an IdP specific wrapper when then calls `sss_radiuskdc_return_padata()` for additional checks.

bye,
Sumit